### PR TITLE
Familiars respawn

### PR DESCRIPTION
**Pull request by Elijahrane**
## About the PR 
As suggested during the discussion yesterday about people griefing the chaplain, familiars respawn now, which makes random deaths to atmos or griefers a bit less devastating to their round. 3 minutes for Remilia, 5 minutes for Cerberus.

**Changelog**

:cl: Rane
- add: The chaplain's familiars can be resummoned now after they die. 3 minutes for Remilia, 5 minutes for Cerberus.

### CODE REVIEW

File: Content.Server/Bible/BibleSystem.cs [hunk 1]
```diff
@@ -35,6 +34,35 @@ public override void Initialize()
...
+            SubscribeLocalEvent<FamiliarComponent, MobStateChangedEvent>(OnFamiliarDeath);
+        }
+        /// <summary>
+        /// This handles familiar respawning.
+        /// </summary>
+        public override void Update(float frameTime)
+        {
+            base.Update(frameTime);
+
+            foreach (var (summonableComp, respawning) in EntityQuery<SummonableComponent, SummonableRespawningComponent>())
```
Comment (metalgearsloth) [inline Content.Server/Bible/BibleSystem.cs:46]: Least common component first.

File: Content.Server/Bible/BibleSystem.cs [hunk 2]
```diff
@@ -35,6 +34,35 @@ public override void Initialize()
...
+                {
+                    EntityManager.DeleteEntity(summonableComp.Summon.Value);
+                    summonableComp.Summon = null;
+                }
+                summonableComp.AlreadySummoned = false;
+                _popupSystem.PopupEntity(Loc.GetString("bible-summon-respawn-ready", ("book", summonableComp.Owner)), summonableComp.Owner, Filter.Pvs(summonableComp.Owner));
+                SoundSystem.Play(Filter.Pvs(summonableComp.Owner), "/Audio/Effects/radpulse9.ogg", summonableComp.Owner, AudioParams.Default.WithVolume(-4f));
+                /// Clean up the accumulator and respawn tracking component
+                summonableComp.Accumulator = 0;
+                RemComp<SummonableRespawningComponent>(summonableComp.Owner);
```
Comment (metalgearsloth) [inline Content.Server/Bible/BibleSystem.cs:64]: Isn't this going to have collection modified issues doing it in the loop.

File: Content.Server/Bible/Components/SummonableComponent.cs [hunk 3]
```diff
@@ -30,5 +35,12 @@ public sealed class SummonableComponent : Component
             Description = "bible-summon-verb-desc",
             Event = new SummonActionEvent(),
         };
+
+        /// Used for respawning
+        [ViewVariables]
+        public float Accumulator = 0f;
```
Comment (metalgearsloth) [inline Content.Server/Bible/Components/SummonableComponent.cs:41]: Needs to be a datafield to be serialized correctly.

File: Content.Server/Bible/BibleSystem.cs [hunk 4]
```diff
@@ -125,6 +154,24 @@ private void OnSummon(EntityUid uid, SummonableComponent component, SummonAction
...
+        /// </summary>
+        private void OnFamiliarDeath(EntityUid uid, FamiliarComponent component, MobStateChangedEvent args)
+        {
+            if (!args.Component.IsDead() || component.Source == null)
+                return;
+
+            var source = component.Source;
+            if (source != null && TryComp<SummonableComponent>(source, out var summonable))
+            {
+                AddComp<SummonableRespawningComponent>(summonable.Owner);
```
Comment (metalgearsloth) [inline Content.Server/Bible/BibleSystem.cs:170]: Should this be EnsureComp instead.

**Review by metalgearsloth — CHANGES_REQUESTED** [review]

**Comment by Elijahrane** [comment]
Very slightly out of scope but I made all the other accumulators in components datafields too while I was at it

File: Content.Server/Bible/BibleSystem.cs [hunk 5]
```diff
@@ -138,7 +201,17 @@ private void AttemptSummon(SummonableComponent component, EntityUid user, Transf
...
-            EntityManager.SpawnEntity(component.SpecialItemPrototype, position.Coordinates);
+            // Make this familiar the component's summon
+            var familiar = EntityManager.SpawnEntity(component.SpecialItemPrototype, position.Coordinates);
+                            component.Summon = familiar;
+
+            /// We only want to add the familiar component to mobs
+            if (HasComp<MobStateComponent>(familiar))
+            {
+                /// Make this Summon the familiar's source
+                var familiarComp = AddComp<FamiliarComponent>(familiar);
```
Comment (mirrorcult) [inline Content.Server/Bible/BibleSystem.cs:212]:
```suggestion
                var familiarComp = EnsureComp<FamiliarComponent>(familiar);
```

**Review by mirrorcult — APPROVED** [review]

**Comment by Elijahrane** [comment]
Reviews addressed

**Pull request merged**